Red-Black Tree/RedBlackTree.playground/Sources/RedBlackTree.swift

```diff
@@ -0,0 +1,751 @@
+    }
+  }
+  
+  var uncle: RBNode? {
+    return parent?.sibling
+  }
+}
+
+// MARK: - RedBlackTree
+
+public class RedBlackTree<Key: Comparable> {
+  public typealias RBNode = RBTreeNode<Key>
+  
+  fileprivate(set) var root: RBNode
+  fileprivate(set) var size = 0
+  fileprivate let nullLeaf = RBNode()
+  
+  public init() {
+    root = nullLeaf
+  }
+}
+
+// MARK: - Equatable protocol
+
+extension RBTreeNode {
+  static public func == <T>(lhs: RBTreeNode<T>, rhs: RBTreeNode<T>) -> Bool {
+    return lhs.key == rhs.key
+  }
+}
+
+// MARK: - Finding a nodes successor
+
+extension RBTreeNode {
+  /*
+   * Returns the inorder successor node of a node
+   * The successor is a node with the next larger key value of the current node
+   */
+  public func getSuccessor() -> RBNode? {
+    // If node has right child: successor min of this right tree
+    if let rightChild = self.rightChild {
+      if !rightChild.isNullLeaf {
+        return rightChild.minimum()
+      }
+    }
+    // Else go upward until node left child
+    var currentNode = self
+    var parent = currentNode.parent
+    while currentNode.isRightChild {
+      if let parent = parent {
+        currentNode = parent
+      }
+      parent = currentNode.parent
+    }
+    return parent
+  }
+}
+
+// MARK: - Searching
+
+extension RBTreeNode {
+  /*
+   * Returns the node with the minimum key of the current subtree
+   */
+  public func minimum() -> RBNode? {
+    if let leftChild = leftChild {
+      if !leftChild.isNullLeaf {
+        return leftChild.minimum()
+      }
+      return self
+    }
+    return self
+  }
+  
+  /*
+   * Returns the node with the maximum key of the current subtree
+   */
+  public func maximum() -> RBNode? {
+    if let rightChild = rightChild {
+      if !rightChild.isNullLeaf {
+        return rightChild.maximum()
+      }
+      return self
+    }
+    return self
+  }
+}
+
+extension RedBlackTree {
+  /*
+   * Returns the node with the given key |input| if existing
+   */
+  public func search(input: Key) -> RBNode? {
+    return search(key: input, node: root)
+  }
+  
+  /*
+   * Returns the node with given |key| in subtree of |node|
+   */
+  fileprivate func search(key: Key, node: RBNode?) -> RBNode? {
+    // If node nil -> key not found
+    guard let node = node else {
+      return nil
+    }
+    // If node is nullLeaf == semantically same as if nil
+    if !node.isNullLeaf {
+      if let nodeKey = node.key {
+        // Node found
+        if key == nodeKey {
+          return node
+        } else if key < nodeKey {
+          return search(key: key, node: node.leftChild)
+        } else {
+          return search(key: key, node: node.rightChild)
+        }
+      }
+    }
+    return nil
+  }
+}
+
+// MARK: - Finding maximum and minimum value
+
+extension RedBlackTree {
+  /*
+   * Returns the minimum key value of the whole tree
+   */
+  public func minValue() -> Key? {
+    if let minNode = root.minimum() {
```

Comment:
this is fine, something a little cleaner would be to use a guard let statement. 

E.g. 
```
guard let minNode = root.minimum()  else { return nil }
return minNode.key
```